source/declarations.tex

```diff
@@ -248,14 +248,33 @@
 is contextually converted to \keyword{bool} and
 the converted expression shall be a constant expression\iref{expr.const}.
 If the value of the expression when
-so converted is \tcode{true}, the declaration has no
-effect. Otherwise, the program is ill-formed, and the resulting
+so converted is \tcode{true}
+or the expression is evaluated in the context of a template definition,
+the declaration has no
+effect. Otherwise,
+%FIXME: What does it mean for a grammar term to fail? Also in [temp.res]/p6.
+the \grammarterm{static_assert-declaration} fails,
+the program is ill-formed, and the resulting
 diagnostic message\iref{intro.compliance} should include the text of
 the \grammarterm{string-literal}, if one is supplied.
 \begin{example}
 \begin{codeblock}
 static_assert(sizeof(int) == sizeof(void*), "wrong pointer size");
 static_assert(sizeof(int[2]));          // OK, narrowing allowed
+
+template <class T>
+void f(T t) {
+  if constexpr (sizeof(T) == sizeof(int)) {
+    use(t);
+  } else {
+    static_assert(false, "must be int-sized");
+  }
+}
+
+void g(char c) {
+  f(0);                                 // OK
+  f(c);                                 // error: must be int-sized
```

Comment:
What's the take on `error:` comments that can be OK depending on the parameters of the abstract machine? This particular one is called out by
> [2](https://eel.is/c++draft/intro.abstract#2)
[#](http://github.com/Eelis/draft/tree/7eaad29c229ece4591b805688a9827b25d2149bf/source/intro.tex#L900)
Certain aspects and operations of the abstract machine are described in this document as implementation-defined (for example, sizeof(int))[.](https://eel.is/c++draft/intro.abstract#2.sentence-1)
> -- https://eel.is/c++draft/intro.abstract#2.sentence-1

Reply:
Yes, that's a problem with the example, since `sizeof(char) == sizeof(int)` is possible.
Maybe change the comment to read "error on implementations where sizeof(int) > 1: must be int-sized".

Reply:
I'm positive there's a precedent here.

Reply:
Nice work, thanks!